auraed/src/runtime/cell_service/cells/cells.rs

```diff
@@ -46,24 +46,30 @@ pub struct Cells {
 // - Get Cgroup and pids from executable_name
 
 impl Cells {
-    /// Add the [Cell] to the cache with key [CellName].
-    /// Returns an error if a duplicate [CellName] already exists in the cache.
+    /// Calls [Cell::allocate] on a new [Cell] and adds it to it's cache with key [CellName].
+    ///
+    /// # Errors
+    /// * If cell exits -> [CellsError::CellExists]
+    /// * If a cell is not in cache but cgroup exists on fs -> [CellsError::CgroupIsNotACell]
+    /// * If cell fails to allocate (see [Cell::allocate])
     pub fn allocate(
         &mut self,
         cell_name: CellName,
         cell_spec: CellSpec,
     ) -> Result<&Cell> {
-        // TODO: replace with this when it becomes stable
-        // cache.try_insert(cell_name.clone(), cgroup)
-
-        // Check if there was already a cgroup in the table with this cell name as a key.
-        if self.cache.contains_key(&cell_name) {
-            // TODO Reconcile cache against filesystem somewhere, maybe here?
-            // TODO https://github.com/aurae-runtime/aurae/issues/198
-            return Err(CellsError::CellExists { cell_name });
+        if Cgroup::exists(&cell_name) {
+            return if self.cache.contains_key(&cell_name) {
+                Err(CellsError::CellExists { cell_name })
+            } else {
+                Err(CellsError::CgroupIsNotACell {
+                    cell_name: cell_name.clone(),
+                })
+            };
         }
 
-        // `or_insert` will always insert as we've already assured ourselves that the key does not exist.
+        // From here, we know the cgroup doesn't exist, so remove from cache if it does
+        let _ = self.cache.remove(&cell_name);
```

Comment:
should we warn or something? aiui this is user error: they removed the cgroup from the FS without calling `free` on auraed.

Reply:
Barring logic errors in the code, I think this has to be user error as well. The user must have used aurae to create the cell and then somehow the cgroup is no longer on the host.

I don't know the correct response, but I put a `warn!` for now and a TODO (happy to change the logic in this PR if we want to make a decision)

> // TODO: Should we not remove the cell (that has no cgroup) from the cache and
    //       force the user to call Free? Free will also return an error, but we may be
    //       calling other logic in free that we want to run.

I also made cell's drop call `free` (not just in tests).